### ISSUE TITLE
table_exists runs a query in the database

### ISSUE THREAD

**Issue by edublancas** [ploomber/jupysql#365]
I noticed that this implementation: https://github.com/ploomber/jupysql/blob/d721d56d0975d3a6e78dce366a0c5bfa1deeead1/src/sql/util.py#L162

runs a query even when the user passes the `with` argument. This is inefficient since the with arguments will lead to building a CTE that might take a lot to run. Do we actually need this?

Isn't checking `table_name in store` enough? We can assume that if a user ran a `--save table_name` argument, then running `--with table_name` won't lead to any "table not found error" since the declaration exists (it's not a table but it's a CTE). Am I missing anything?

There is another side effect: if the definition isn't valid SQL.

```
%%sql --save table_name --no-execute
not valid SQL
```

If we ran the cell above, table_exists will return false because the query is wrapper in a try, catch statement, which will cause the "table_name does not exist" error to be displayed, which is misleading because the error is that the SQL code isn't valid, not that the table doesn't exist.

I feel like `table_name in store` should be enough, but perhaps I'm missing something here @yafimvo

**Comment by yafimvo** [comment]
Yes, good point. Adding this will make it more efficient and notify the user with a better error message.
I added this logic to this [PR](https://github.com/ploomber/jupysql/pull/361). 

Since now, if the table exists and the correct `--with` was provided, we should be able to reproduce the error from this [issue](https://github.com/ploomber/jupysql/issues/355). If we remove this line  `-- remove top 1% outliers for better visualization` we should be able to get the plot, however, it fails with a different error which I'm not sure is related to these changes.


## Reproduce

Running these 2 code cells:

```
%%sql --save many_passengers --no-execute
SELECT *
FROM taxi
WHERE passenger_count > 3
-- remove top 1% outliers for better visualization
AND trip_distance < 18.93
```

```
%sqlplot histogram --table many_passengers --column trip_distance --with many_passengers
```

Should return this TypeError (works OK)

```
TypeError: %o format: an integer is required, not dict
```

But running these cells should return a plot without any errors

```
%%sql --save many_passengers --no-execute
SELECT *
FROM taxi
WHERE passenger_count > 3
AND trip_distance < 18.93
```
```
%%sql --save many_passengers --no-execute
SELECT *
FROM taxi
WHERE passenger_count > 3
AND trip_distance < 18.93
```

But it returns this error:

```
ProgrammingError: (pymysql.err.ProgrammingError) (1064, 'You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near \'"many_passengers"\' at line 9')
[SQL: WITH `many_passengers` AS (
SELECT *
FROM taxi
WHERE passenger_count > 3
AND trip_distance < 18.93)
SELECT
    MIN("trip_distance"),
    MAX("trip_distance")
FROM "many_passengers"]
(Background on this error at: https://sqlalche.me/e/14/f405)
```


I think it is something with the query format we try to run:

This was the previous format (working)

```
%%sql
WITH `many_passengers` AS (
SELECT *
FROM taxi
WHERE passenger_count > 3
AND trip_distance < 18.93)
SELECT
    MIN("trip_distance"),
    MAX("trip_distance")
FROM "many_passengers"
```

This is the current one (failing)

```
%%sql
WITH many_passengers AS (
    SELECT * 
    FROM taxi 
    WHERE passenger_count > 3 
    AND trip_distance < 18.93) 
SELECT 
MIN(`trip_distance`), 
MAX(`trip_distance`) 
FROM `many_passengers`
```

@edublancas

**Comment by edublancas** [comment]
I fixed the errors you're encountering in https://github.com/ploomber/jupysql/pull/362,

the issue is that having `% o` was interpreted as a parameter interpolation in the SQL driver, so passing a `.text` object to the `.execute` function tells SQLAlchemy to interpret the query as a literal string, instead of a parameterized one.

The other error you described is because some drivers need backticks, so I changed them to use them depending on the dialect. Looks like it fixed all the problems but I'll merge the PR and try to reproduce your findings to see if we're covered.
